format/src/matbin/mod.rs

```diff
@@ -0,0 +1,309 @@
+            next,
+            header.sampler_count.get() as usize,
+        )?;
+
+        Some(Self {
+            bytes,
+            header: header.into_ref(),
+            parameters,
+            samplers,
+        })
+    }
+
+    pub fn shader_path(&self) -> Result<Cow<'_, U16Str>, MatbinError> {
+        let offset = self.header.shader_path_offset.get() as usize;
+        let bytes = &self.bytes[offset..];
+
+        Ok(read_utf16_string(bytes)?)
+    }
+
+    pub fn source_path(&self) -> Result<Cow<'_, U16Str>, MatbinError> {
+        let offset = self.header.source_path_offset.get() as usize;
+        let bytes = &self.bytes[offset..];
+
+        Ok(read_utf16_string(bytes)?)
+    }
+
+    pub fn samplers(
+        &self,
+    ) -> impl Iterator<Item=Result<SamplerIterElement, MatbinError>> {
+        self.samplers.iter()
+            .map(|e| {
+                let sampler_type = {
+                    let offset = e.type_offset.get() as usize;
+                    let bytes = &self.bytes[offset..];
+                    read_utf16_string(bytes)
+                }?;
+
+                let path = {
+                    let offset = e.path_offset.get() as usize;
+                    let bytes = &self.bytes[offset..];
+                    read_utf16_string(bytes)
+                }?;
+
+                Ok(SamplerIterElement {
+                    sampler_type,
+                    path,
+                })
+            })
+    }
+
+    pub fn parameters(
+        &self,
+    ) -> impl Iterator<Item=Result<ParameterIterElement, MatbinError>> {
+        self.parameters.iter()
+            .map(|e| {
+                let name = {
+                    let offset = e.name_offset.get() as usize;
+                    let bytes = &self.bytes[offset..];
+                    read_utf16_string(bytes)
+                }?;
+
+                let value_slice = &self.bytes[e.value_offset.get() as usize..];
+                let value = ParameterValue::from_type_and_slice(
+                    e.value_type.get(),
+                    value_slice,
+                )?;
+
+                Ok(ParameterIterElement {
+                    name,
+                    value,
+                })
+            })
+    }
+}
+
+pub struct ParameterIterElement<'buffer> {
+    pub name: Cow<'buffer, U16Str>,
+    pub value: ParameterValue<'buffer>,
+}
+
+pub struct SamplerIterElement<'buffer> {
+    pub sampler_type: Cow<'buffer, U16Str>,
+    pub path: Cow<'buffer, U16Str>,
+}
+
+impl<'buffer> std::fmt::Debug for Matbin<'buffer> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Matbin")
+            .field("shader_path", &self.shader_path())
+            .field("source_path", &self.source_path())
+            .field("header", self.header)
+            .field("parameters", &self.parameters)
+            .field("samplers", &self.samplers)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
```

Comment:
A custom impl of `Debug` would be better here, the `Debug` impl for F32<O> etc. isn't great.